File: python/gameplay_policy/piece.py
from dataclasses import dataclass


@dataclass(frozen=True)
class King:
    color: str


@dataclass(frozen=True)
class Queen:
    color: str


@dataclass(frozen=True)
class Rook:
    color: str


@dataclass(frozen=True)
class Bishop:
    color: str


@dataclass(frozen=True)
class Knight:
    color: str


@dataclass(frozen=True)
class Pawn:
    color: str


Piece = King | Queen | Rook | Bishop | Knight | Pawn

File: java/gameplay-policy/src/main/java/com/hexagonalchess/gameplay/Piece.java
package com.hexagonalchess.gameplay;

public sealed interface Piece permits Piece.King, Piece.Queen, Piece.Rook, Piece.Bishop, Piece.Knight, Piece.Pawn {
    Color color();

    record King(Color color) implements Piece {}
    record Queen(Color color) implements Piece {}
    record Rook(Color color) implements Piece {}
    record Bishop(Color color) implements Piece {}
    record Knight(Color color) implements Piece {}
    record Pawn(Color color) implements Piece {}
}
